src/v/v8_engine/executor.h

```diff
@@ -0,0 +1,210 @@
+        }
+        _on_done.notify();
+    }
+
+    // Get future for seastar thread for wating when executor thread will
+    // complete task
+    ss::future<> get_future() {
+        return _on_done.wait().then([this] {
+            if (_state.failed()) {
+                return ss::make_exception_future<>(
+                  std::move(_state).get_exception());
+            } else {
+                return ss::make_ready_future<>();
+            }
+        });
+    }
+
+    Func _func;
+    ss::future_state<bool> _state;
+    condition _on_done;
+};
+
+// This class implement wrapper for ss::gate for using in ss::sharded
+class local_gate {
+public:
+    ss::future<> stop();
+
+    ss::gate _gate;
+};
+
+// This class implement wrapper for ss::abort_source for using in ss::sharded.
+class local_abort_source {
+public:
+    ss::future<> stop();
+
+    ss::abort_source _abort_source;
+};
+
+// This class implement semaphore with exponential backoff.
+// TODO: mybe use something with fairness guarantee
+class semaphore {
+public:
+    explicit semaphore(size_t free_slots);
+
+    ss::future<> start();
+    ss::future<> stop();
+
+    // Try to acquire semaphore and sleep if semaphore does not contain free
+    // slots
+    ss::future<> acquire();
+
+    // Release sempahore
+    void release();
+
+private:
+    std::atomic<size_t> _free_slots;
+
+    static constexpr ss::lowres_clock::duration _max_wait_time_ms{100};
+
+    // _abort_sources is used for abort sleep if semaphore is stoped
+    ss::sharded<local_abort_source> _abort_sources;
+    std::atomic<bool> _is_closed{false};
+};
+
+} // namespace internal
+
+// This class implement thread pool with std::thread for runing v8 script. It
+// uses boost::lockfree_queue for storing task. Queue task stores #thread_count
+// tasks. Because we need to set watchdog for timeout v8 script only when
+// executor can execute script without wating. semaphore is used for controle
+// does executor have free thread.
+class executor {
+public:
+    executor(size_t threads_num, uint8_t cpu_id);
+
+    executor(const executor& other) = delete;
+    executor& operator=(const executor& other) = delete;
+    executor(executor&& other) = delete;
+    executor& operator=(executor&& other) = delete;
+
+    ~executor();
+
+    // Start ss::sharded for ss::gate and ss::abort_source
+    ss::future<> start();
+
+    // Stop executor. Frst of all we need to stop semaphore for cancel waitnig
+    // tasks. After that we need complete all task in queue and finally stop
+    // std::threads
+    ss::future<> stop();
+
+    bool is_stopping() const;
+
+    /// Submit new task in executor.
+    ///
+    /// \param start_func is used for set watchdog
+    /// \param func_for_executor is v8 script
+    template<typename StartFunc, typename FuncForExecutor>
+    auto submit(StartFunc&& start_func, FuncForExecutor&& func_for_executor) {
+        return ss::with_gate(
```

Comment:
Sometime we use the ss:: API other times we are using c++ atomics / synchronization stuff. We should choose one or the other, IMO

Reply:
But I need to use atomic for semaphore (free_slots) because I can run it from different thread. And I need to understand how many free thread does executor have.

Reply:
main problem is executor is one for all cores

Reply:
Done